src/track/track.cpp

```diff
@@ -328,18 +325,34 @@ bool Track::setBeatsWhileLocked(mixxx::BeatsPointer pBeats) {
     return true;
 }
 
-void Track::setBeatsMarkDirtyAndUnlock(QMutexLocker* pLock, mixxx::BeatsPointer pBeats) {
+bool Track::trySetBeatsMarkDirtyAndUnlock(
+        QMutexLocker* pLock,
+        mixxx::BeatsPointer pBeats,
+        SetBeatsMode mode) {
     DEBUG_ASSERT(pLock);
+
+    if (m_pBeats && m_record.getBpmLocked()) {
+        // Track has already a valid and locked beats object, abbort.
+        qDebug() << "Track beats is already set and BPM-locked. Discard the new beats";
+        return false;
+    }
+
+    if (mode == SetBeatsMode::SetBpmLock) {
+        *(m_record.ptrBpmLocked()) = true;
```

Comment:
The dirty flag has to be set accordingly if any of the two properties has been changed.